### PR TITLE
fix(twap): improve twap orders state synchronisation

### LINKED ISSUE #2632: Improve twap orders state synchronisation
Due to `TwapOrdersUpdater` flaws:
1. A newly created TWAP order doesn't display in the orders table instantly
2. After TWAP order cancellation it can go back to `Pending` state for some time and only then become Cancelled

### PR DESCRIPTION
**Pull request by shoom3301**
# Summary

Fixes #2632

## Cases
1. When creating a new TWAP order, then it should be immediately displayed in the orders table
2. When cancelling a TWAP order, then it and its child should immediately go to `Cancelling` state
3. Once TWAP order is canceled, then it should never change status to `Pending` only to `Cancelled`

# To test (in Limit orders)
1. Orders status displaying in the orders table. Please, especially pay attention on `Cancelling...` and `Cancelled` statuses

### CODE REVIEW

File: src/common/hooks/useCancelOrder/useSendOnChainCancellation.ts [hunk 1]
```diff
@@ -16,29 +17,33 @@ export function useSendOnChainCancellation() {
   const addTransaction = useTransactionAdder()
   const getOnChainCancellation = useGetOnChainCancellation()
 
-  return useCallback(
-    async (order: Order) => {
-      if (!chainId) {
-        return
-      }
+  const processCancelledOrder = useCallback(
```
Comment (shoom3301) [inline src/common/hooks/useCancelOrder/useSendOnChainCancellation.ts:20]: With ComposableCow orders we might cancel more than one order, hence, we need to process it more that once

File: src/modules/ordersTable/pure/OrderStatusBox/index.tsx [hunk 2]
```diff
@@ -60,6 +60,32 @@ const Wrapper = styled.div<{
   }
 `
 
+function getOrderStatusTitle(order: ParsedOrder): string {
```
Comment (shoom3301) [inline src/modules/ordersTable/pure/OrderStatusBox/index.tsx:63]: Actually, there was a comment "Cancelled status takes precedence" but it wasn't true, `isCancelling` had a higher priority. But now everything is in its place.
We need it, because we don't reset `isCancelling` flag for TWAP orders.
Issue to fix it: https://github.com/cowprotocol/cowswap/issues/2634

File: src/modules/twap/hooks/useCreateTwapOrder.ts [hunk 3]
```diff
@@ -32,15 +42,36 @@ export function useCreateTwapOrder() {
...
+    const paramsStruct = buildTwapOrderParamsStruct(chainId, twapOrderWithStartTime)
+    const orderId = getConditionalOrderId(paramsStruct)
 
     tradeConfirmActions.onSign(pendingTrade)
 
     try {
-      const { safeTxHash } = await settleTwapOrder({ ...twapOrder, startTime }, twapOrderCreationContext)
+      const { safeTxHash } = await settleTwapOrder(twapOrderWithStartTime, paramsStruct, twapOrderCreationContext)
+
+      addTwapOrderToList({
```
Comment (shoom3301) [inline src/modules/twap/hooks/useCreateTwapOrder.ts:54]: Immediately add a new twap order to the list

File: src/modules/twap/hooks/useTwapOrdersAuthMulticall.ts [hunk 4]
```diff
@@ -21,12 +21,12 @@ export function useTwapOrdersAuthMulticall(
   const results = useSingleContractMultipleData(composableCowContract, 'singleOrders', input, DEFAULT_LISTENER_OPTIONS)
 
   return useMemo(() => {
-    const loadedResults = results.filter((result) => !result.loading)
+    const loadedResults = results.filter((result) => !result.loading && result.valid && !result.syncing)
```
Comment (shoom3301) [inline src/modules/twap/hooks/useTwapOrdersAuthMulticall.ts:24]: For some reason, the multicall returned a wrong value when `syncing=true`

File: src/modules/ordersTable/pure/OrderStatusBox/index.tsx (hunk 2 above)
Comment (alfetopito) [reply src/modules/ordersTable/pure/OrderStatusBox/index.tsx:63]: That was a bug fix for an edge case.
Can't find out now which one, but anyway. We should test this bit thoroughly to make sure this isn't breaking anything

File: src/modules/twap/updaters/TwapOrdersUpdater.tsx [hunk 5]
```diff
@@ -51,6 +51,10 @@ export function TwapOrdersUpdater(props: {
       const existedOrder = twapOrdersList[info.id]
 
       if (existedOrder) {
+        if (existedOrder.status === TwapOrderStatus.WaitSigning) {
```
Comment (alfetopito) [inline src/modules/twap/updaters/TwapOrdersUpdater.tsx:54]: `existingOrder`?

File: src/modules/twap/utils/updateTwapOrdersList.ts [hunk 6]
```diff
@@ -0,0 +1,17 @@
...
+  const newState = { ...currentState, ...nextState }
+
+  Object.keys(nextState).forEach((orderId) => {
+    const currentOrder = currentState[orderId]
+
+    // Don't update orders in final statuses
+    if (currentOrder && TWAP_NOT_PENDING_STATUSES.includes(currentOrder.status)) {
+      newState[orderId] = currentOrder
+    }
+  })
```
Comment (alfetopito) [inline src/modules/twap/utils/updateTwapOrdersList.ts:14]: I don't get what's happening here

First, the parameter name is `...List` but it's actually an object

Then, both parameters are combined into a single object

Later on line `12` the given order in the new obj is updated with data from `currentState`?

Why do you need to merge both?
Wouldn't it be enough to just iterate over `currentState` keys?

File: src/common/hooks/useCancelOrder/useGetOnChainCancellation.ts [hunk 7]
```diff
@@ -19,13 +19,13 @@ export function useGetOnChainCancellation(): (order: Order) => Promise<OnChainCa
 
   return useCallback(
     (order: Order) => {
-      const composableCowId = order.composableCowInfo?.id
+      const isComposableCowOrder = !!order.composableCowInfo?.id
```
[the remark this answers is not in the document]
Comment (shoom3301) [reply src/common/hooks/useCancelOrder/useGetOnChainCancellation.ts:22]: Great suggestion! Added, thanks!

File: src/modules/twap/updaters/TwapOrdersUpdater.tsx (hunk 5 above)
Comment (shoom3301) [reply src/modules/twap/updaters/TwapOrdersUpdater.tsx:54]: Thanks, fixed

File: src/modules/twap/utils/updateTwapOrdersList.ts (hunk 6 above)
Comment (shoom3301) [reply src/modules/twap/utils/updateTwapOrdersList.ts]: Agree, this code is odd.
Refactored it and added comments:
https://github.com/cowprotocol/cowswap/pull/2633/commits/c71508ac09e854eebbd8b1653aeffa54ec8816d7

File: src/modules/ordersTable/pure/OrderStatusBox/index.tsx (hunk 2 above)
Comment (shoom3301) [reply src/modules/ordersTable/pure/OrderStatusBox/index.tsx:63]: Thank you! 
Added test instructions for @elena-zh

File: src/modules/twap/utils/updateTwapOrdersList.ts [hunk 8]
```diff
@@ -0,0 +1,24 @@
...
+ * It only updates an order if it in a pending state
+ */
+export function updateTwapOrdersList(currentState: TwapOrdersList, nextState: TwapOrdersList): TwapOrdersList {
+  // Filter nextState and left only orders with pending statuses
+  const newState = Object.keys(nextState).reduce<TwapOrdersList>((acc, orderId) => {
+    const currentOrder = currentState[orderId]
+
+    // Update an order only if it's in pending state
+    // Then don't update it
+    if (!currentOrder || TWAP_PENDING_STATUSES.includes(currentOrder.status)) {
```
Comment (alfetopito) [inline src/modules/twap/utils/updateTwapOrdersList.ts:16]: Shouldn't you do this if `currentOrder` exists?

The comment only mentions if it's in a pending state, but there's no mention about the currentState order

```suggestion
    if (currentOrder || TWAP_PENDING_STATUSES.includes(currentOrder.status)) {
```

File: src/utils/orderUtils/getIsComposableCowOrder.ts [hunk 9]
```diff
@@ -0,0 +1,5 @@
+import { ComposableCowInfo } from 'common/types'
+
+export function getIsComposableCowOrder(order: { composableCowInfo?: ComposableCowInfo }): boolean {
```
Comment (alfetopito) [inline src/utils/orderUtils/getIsComposableCowOrder.ts:3]: More explicit?

```suggestion
export function getIsComposableCowParentOrder(order: { composableCowInfo?: ComposableCowInfo }): boolean {
```

File: src/modules/twap/utils/updateTwapOrdersList.ts (hunk 8 above)
Comment (alfetopito) [reply src/modules/twap/utils/updateTwapOrdersList.ts:17]: Ahhh ok, now I got it!
